parachain/pallets/ethereum-beacon-client/src/lib.rs

```diff
@@ -780,5 +771,102 @@ pub mod pallet {
 
 			Ok(sync_committee)
 		}
+
+		// Verifies that the receipt encoded in proof.data is included
+		// in the block given by proof.block_hash. Inclusion is only
+		// recognized if the block has been finalized.
+		fn verify_receipt_inclusion(proof: &Proof) -> Result<Receipt, DispatchError> {
+			let stored_header =
+				<ExecutionHeaders<T>>::get(proof.block_hash).ok_or(Error::<T>::MissingHeader)?;
+
+			let result = stored_header
+				.check_receipt_proof(&proof.data.1)
+				.ok_or(Error::<T>::InvalidProof)?;
+
+			match result {
+				Ok(receipt) => Ok(receipt),
+				Err(err) => {
+					log::trace!(
+						target: "ethereum-beacon-client",
+						"Failed to decode transaction receipt: {}",
+						err
+					);
+					Err(Error::<T>::InvalidProof.into())
+				},
+			}
+		}
+	}
+
+	impl<T: Config> Verifier for Pallet<T> {
+		/// Verify a message by verifying the existence of the corresponding
+		/// Ethereum log in a block. Returns the log if successful.
+		fn verify(message: &Message) -> Result<Log, DispatchError> {
+			log::info!(
+				target: "ethereum-beacon-client",
+				"💫 Verifying message with block hash {}",
+				message.proof.block_hash,
+			);
+
+			let receipt_result = Self::verify_receipt_inclusion(&message.proof);
+			if let Err(receipt_result) = &receipt_result {
+				log::trace!(
```

Comment:
I added more trace logging for errors so that we can debug issues if any arise, especially because this is the newest part of the Eth2 work.